modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java

```diff
@@ -0,0 +1,487 @@
+
+            fut.complete(desc);
+        }
+
+        return fut.thenApply(res -> {
+            registriesVv.removeWhenComplete(schemaListener);
+            return res;
+        });
+    }
+
+    /**
+     * Checks that the provided schema version is less or equal than the latest version from the schema registry.
+     *
+     * @param tblId Unique table id.
+     * @param schemaVer Schema version for the table.
+     * @return True, if the schema version is less or equal than the latest version from the schema registry, false otherwise.
+     */
+    private boolean checkSchemaVersion(int tblId, int schemaVer) {
+        SchemaRegistry registry = registriesVv.latest().get(tblId);
+
+        assert registry != null : IgniteStringFormatter.format("Registry for the table not found [tblId={}]", tblId);
+
+        return schemaVer <= registry.lastSchemaVersion();
+    }
+
+    /**
+     * Try to find schema in cache.
+     *
+     * @param tblId Table id.
+     * @param schemaVer Schema version.
+     * @return Descriptor if required schema found, or {@code null} otherwise.
+     */
+    private @Nullable SchemaDescriptor searchSchemaByVersion(int tblId, int schemaVer) {
+        SchemaRegistry registry = registriesVv.latest().get(tblId);
+
+        if (registry != null && schemaVer <= registry.lastSchemaVersion()) {
+            return registry.schema(schemaVer);
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get the schema registry for the given causality token and table id.
+     *
+     * @param causalityToken Causality token.
+     * @param tableId Id of a table which the required registry belongs to. If {@code null}, then this method will return a future which
+     *     will be completed with {@code null} result, but only when the schema manager will have consistent state regarding given causality
+     *     token.
+     * @return A future which will be completed when schema registries for given causality token are ready.
+     */
+    public CompletableFuture<SchemaRegistry> schemaRegistry(long causalityToken, int tableId) {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteException(NODE_STOPPING_ERR, new NodeStoppingException());
+        }
+
+        try {
+            return registriesVv.get(causalityToken)
+                    .thenApply(regs -> inBusyLock(busyLock, () -> regs.get(tableId)));
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Returns schema registry by table id.
+     *
+     * @param tableId Table id.
+     * @return Schema registry.
+     */
+    public SchemaRegistry schemaRegistry(int tableId) {
+        return registriesVv.latest().get(tableId);
+    }
+
+    /**
+     * Drop schema registry for the given table id.
+     *
+     * @param causalityToken Causality token.
+     * @param tableId Table id.
+     */
+    public CompletableFuture<?> dropRegistry(long causalityToken, int tableId) {
```

Comment:
Seems, we also should drop history from metastorage for the given table along with its registry.